#Created on 3/10/2015
__author__ = 'Juan A. Ugalde'

if __name__ == '__main__':
    import os
    import sys
    import argparse
    from tools.data_input.GenomeData import read_genome_list
    from tools.data_input.ClusterInput import parse_fastortho, parse_tribemcl

    #Create the options and program description
    program_description = "This script evaluates if from a list of clusters, how many genomes from another list are" \
                          "present in those clusters. The idea is to use this to look for cluster that are either" \
                          "under or over represented in a group of genomes." \
                          " The inputs are:" \
                          "-l genome list. The list of genomes to evaluate with the clusters."\
                          "-c cluster file. Original cluster file generated by either TribeMCL or OrthoMCL"\


                          "-List of clusters, generated by orthoMCL" \
                          "-A genome list" \
                          "-A folder with fasta files" \
                          "- An optional group file, to group genomes. For example, all genomes from the same species "

    parser = argparse.ArgumentParser(description=program_description)

    parser.add_argument("-l", "--genome_list_index", type=str,
                        help="File with the genome list. Format GenomeID, FullName, ShortName", required=True)
    parser.add_argument("-c", "--cluster_file", type=str, help="Ortholog file", required=True)
    parser.add_argument("-f", "--fasta_aa_directory", type=str, help="Directory with the fasta files", required=True)
    parser.add_argument("-t", "--clustering_algorithm", type=str, help="Type of clustering used. Options are fastortho,"
                                                                       "tribemcl", required=True)
    parser.add_argument("-g", "--group_information", type=str, help="Group file")
    parser.add_argument("-o", "--output_directory", type=str, help="Output directory", required=True)

    args = parser.parse_args()